### PR TITLE
Cache most frequent queries

### PR DESCRIPTION
**Pull request by juliushaertl**
Contributes to reducing the amount of sql queries that are issued when setting up the filesystem. 

This PR is based on a rebased version of https://github.com/nextcloud/deck/pull/3365 (needs another cleanup after the merge)

- First commit addesses the most frequent query that would resolve the relation from the card id to the board id. This one is now cached in a distributed memory cache. The only condition that the cache should become invalid is when a card is moved to a different board where it is invalidated manually, therefore no TTL is used. (44042 queries in todays query log)
- Second commit aggregates some more savings when fetching board details of single boards (and their labels or ACL). The ACL were the most frequent ones (19111 queries in todays log) but this will would also help in other places where board details may be obtained multiple times in the same request.

### CODE REVIEW

File: lib/Db/CardMapper.php [hunk 1]
```diff
@@ -547,11 +557,17 @@ public function isOwner($userId, $cardId): bool {
...
 	public function findBoardId($id): ?int {
-		$sql = 'SELECT id FROM `*PREFIX*deck_boards` WHERE `id` IN (SELECT board_id FROM `*PREFIX*deck_stacks` WHERE id IN (SELECT stack_id FROM `*PREFIX*deck_cards` WHERE id = ?))';
-		$stmt = $this->db->prepare($sql);
-		$stmt->bindParam(1, $id, \PDO::PARAM_INT);
-		$stmt->execute();
-		return $stmt->fetchColumn() ?? null;
+		//$this->cache->remove('findBoardId:' . $id);
+		$result = $this->cache->get('findBoardId:' . $id);
+		if ($result === null) {
+			$sql = 'SELECT id FROM `*PREFIX*deck_boards` WHERE `id` IN (SELECT board_id FROM `*PREFIX*deck_stacks` WHERE id IN (SELECT stack_id FROM `*PREFIX*deck_cards` WHERE id = ?))';
```
Comment (nickvergessen) [inline lib/Db/CardMapper.php:563]: btw you can put this in query builders as some point:
https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/private/Repair/RemoveLinkShares.php#L115

**Review by nickvergessen — APPROVED** [review]

File: lib/Db/CardMapper.php [hunk 2]
```diff
@@ -107,6 +113,10 @@ public function update(Entity $entity, $updateModified = true): Entity {
 			} catch (Exception $e) {
 			}
 		}
+		// Invalidate cache when the card may be moved to a different board
+		if ($updatedFields['stackId']) {
+			$this->cache->remove('findBoardId:' . $entity->getId());
```
Comment (juliushaertl) [inline lib/Db/CardMapper.php:118]: - [ ] Probably also makes sense to invalidate on delete

File: lib/Db/BoardMapper.php [hunk 3]
```diff
@@ -42,7 +42,10 @@ class BoardMapper extends QBMapper implements IPermissionMapper {
 
 	private $circlesEnabled;
 
+	/** @var CappedMemoryCache */
```
Comment (juliushaertl) [inline lib/Db/BoardMapper.php:45]: - [ ] reuse boardCache and only store board ids in the userBoardCache

File: lib/Db/CardMapper.php (hunk 1 above)
Comment (juliushaertl) [reply lib/Db/CardMapper.php:563]: @nickvergessen Will do here right away then. Thinking about the query, weren't joins usually also a bit faster than sub-queries?

Comment (nickvergessen) [reply lib/Db/CardMapper.php:563]: Yeah, looking at this query if there is always a 1:n (card to stack and stack to board) relation, you could optimise for reading and add a board id column to the cards table as well and then save 2 queries?

Comment (juliushaertl) [reply lib/Db/CardMapper.php]: Moved to a single join query for now as introducing a dedicated board_id column would be a larger topic with possibly heavy migrations involved depending on the table size, should already be a good improvement now with the additional caching in place.

File: lib/Db/BoardMapper.php (hunk 3 above)
Comment (juliushaertl) [reply lib/Db/BoardMapper.php:45]: Thinking more about that it should be fine as it is, the current way has some memory overhead but we can avoid juggling around the existing objects to map to board ids and the other way round when obtaining the boards from the cache.

**Comment by stefan-niedermann** [comment]
Since i saw adjusted unit tests (changing properties from `[]` to `null`) i am wondering whether those changes are breaking API changes? cc @desperateCoder

**Comment by juliushaertl** [comment]
Good catch, yes, though the fetching of acls/labels should happen and they should then be set with an empty array for the API requests that expose a Board entity this is worth double checking (or making sure that we stick to the current empty array in any case when performing the json serialization).

**Review by mejo- — APPROVED** [review]
Overall looks good to me. I went through the code and didn't spot anything obvious :)

**Review by eneiluj — APPROVED** [review]
L:eyes:ks good.

**Comment by stefan-niedermann** [comment]
Maybe i did miss it while reviewing the code, but it looks like the `ETag`s are not updated in the cache? This would lead to more trouble with outdated data (like #3183, #2892 or #2866)

**Review by stefan-niedermann — CHANGES_REQUESTED** [review]
Maybe i did miss it while reviewing the code, but it looks like the `ETag`s are not updated in the cache? This would lead to more trouble with outdated data (like #3183, #2892 or #2866)

**Comment by juliushaertl** [comment]
> Maybe i did miss it while reviewing the code, but it looks like the `ETag`s are not updated in the cache? This would lead to more trouble with outdated data (like #3183, #2892 or #2866)

The board caching layer is a CappedMemoryCache which means it is only persisted during the same request. Meaning any request will at least do one fetch from the database and would then get the latest etag for the board.

**Review by stefan-niedermann — APPROVED** [review]

**Comment by juliushaertl** [comment]
Should also get an adapted backport to 21/22.

**Comment by juliushaertl** [comment]
Manual backport in https://github.com/nextcloud/deck/pull/3740

**Pull request merged**